### PR TITLE
multimodality example

### PR DESCRIPTION
**Pull request by taltlusty**
please review , thanks

### CODE REVIEW

File: fuse_examples/classification/multimodality/loss_multimodal_contrastive_learning.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        label_vec = torch.unsqueeze(label, 0)
+        mask = torch.eq(torch.transpose(label_vec, 0, 1), label_vec).float()
+        logits_imaging_tabular = torch.matmul(imaging_representations, torch.transpose(tabular_representations, 0, 1))/self.temperature
+        logits_tabular_imaging = torch.matmul(tabular_representations, torch.transpose(imaging_representations, 0, 1))/self.temperature
+        loss_imaging_tabular = softcrossentropyloss(mask, logits_imaging_tabular)/torch.sum(mask, 0)
+        loss_tabular_imaging = softcrossentropyloss(mask, logits_tabular_imaging)/torch.sum(mask, 0)
+        return self.alpha*loss_tabular_imaging.sum() + (1-self.alpha)*loss_imaging_tabular.sum()
+
+
+if __name__ == '__main__':
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/loss_multimodal_contrastive_learning.py:51]: remove main

File: fuse_examples/classification/multimodality/mg_dataset.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+
+from fuse_examples.classification.MG_CMMD.input_processor import FuseMGInputProcessor
+from fuse.data.processor.processor_dataframe import FuseProcessorDataFrame
+
+
+from typing import Dict, List
+import torch
+from fuse.utils.utils_hierarchical_dict import FuseUtilsHierarchicalDict
+
+class PostProcessing:
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/mg_dataset.py:25]: Add in comments what this post-processor does.

File: fuse_examples/classification/multimodality/mg_dataset.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        self.imaging_features_lst = imaging_features_lst
+        self.non_imaging_features_lst = non_imaging_features_lst
+        self.use_imaging = use_imaging
+        self.use_non_imaging = use_non_imaging
+
+    def __call__(self, batch_dict: Dict) -> Dict:
+        if not self.use_imaging and not self.use_non_imaging:
+            raise ValueError('No features are in use')
+        mask_list = self.use_imaging * self.imaging_features_lst + self.use_non_imaging * self.non_imaging_features_lst
+        mask_continuous = torch.zeros(len( self.continuous_tabular_features_lst))
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/mg_dataset.py:40]: each feature is necessarily a single value? could it include 3 values like spacing which includes value per axis?

File: fuse_examples/classification/multimodality/mg_dataset.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        FuseUtilsHierarchicalDict.set(batch_dict, 'data.label', label.long())
+        feature_lst = self.continuous_tabular_features_lst + self.categorical_tabular_features_lst + self.label_lst
+        for feature in feature_lst:
+            FuseUtilsHierarchicalDict.pop(batch_dict, 'data.' + feature)
+        return batch_dict
+
+
+# feature selection univarient analysis
+
+def get_selected_features_clinical(output_path):
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/mg_dataset.py:78]: Are you using the code from this line forward? seems like pre-dataset creation code.

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+from fuse.models.backbones.backbone_mlp import FuseMultilayerPerceptronBackbone
+from fuse.utils.utils_hierarchical_dict import FuseUtilsHierarchicalDict
+from typing import Dict, Tuple, Sequence
+from fuse.models.backbones.backbone_inception_resnet_v2 import FuseBackboneInceptionResnetV2
+
+
+
+# class Fusefchead(nn.Module):
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:11]: remove if not used.

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+class TabularImagingConcat(nn.Module):
+    def __init__(self, pooling='max',projection_imaging: nn.Conv2d = nn.Conv2d(384, 256, kernel_size=1, stride=1)):
+        super().__init__()
+        assert pooling in ('max', 'avg')
+        self.pooling = pooling
+        self.projection_imaging = projection_imaging
+
+    def fix_imaging(self,imaging_features):
+        if self.pooling == 'max':
+            imaging_features = F.max_pool2d(imaging_features, kernel_size=imaging_features.shape[2:])
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:124]: You assume 2D here, maybe we can support 3D as well.

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+#         imaging_representations = torch.squeeze(imaging_representations)
+#         FuseUtilsHierarchicalDict.set(batch_dict, 'model.imaging_representations', imaging_representations)
+#
+#         return batch_dict['model']
+
+
+
+# concat model
+class TabularImagingConcat(nn.Module):
+    def __init__(self, pooling='max',projection_imaging: nn.Conv2d = nn.Conv2d(384, 256, kernel_size=1, stride=1)):
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:116]: About projection_imaging:
1. Don't create layers/ models as default argument values - they will be shared with all instances of this class.
2. The type should be more general (nn.Module).

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        super().__init__()
+        assert pooling in ('max', 'avg')
+        self.pooling = pooling
+        self.projection_imaging = projection_imaging
+
+    def fix_imaging(self,imaging_features):
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:122]: This all function can be factor out to projection_imaging module.

File: fuse_examples/classification/multimodality/model_tabular_imaging.py (hunk 7 above)
Comment (mosheraboh) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:116]: What about projection clinical features?

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+        imaging_features = FuseUtilsHierarchicalDict.get(batch_dict, 'model.imaging_features')
+        tabular_features = FuseUtilsHierarchicalDict.get(batch_dict, 'model.tabular_features')
+        imaging_features = self.fix_imaging(imaging_features)
+        res = torch.cat([tabular_features, imaging_features], dim=1)
+        return res
+
+
+#Tabular model
+class FuseModelTabularContinuousCategorical(torch.nn.Module):
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:144]: FYI
In fuse2, I removed the "Fuse" prefix

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    def __init__(self,
+                 continuous_tabular_input: Tuple[Tuple[str, int], ...],
+                 categorical_tabular_input: Tuple[Tuple[str, int], ...],
+                 backbone_categorical_tabular: FuseMultilayerPerceptronBackbone,
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:148]: Does it have to be FuseMultilayerPerceptronBackbone? can it be a different backbone?

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+                 backbone_continuous_tabular: FuseMultilayerPerceptronBackbone,
+                 heads: Sequence[torch.nn.Module],
+                 ) -> None:
+        super().__init__()
+        self.continuous_tabular_input = continuous_tabular_input
+        self.categorical_tabular_input = categorical_tabular_input
+        self.backbone_categorical_tabular = backbone_categorical_tabular
+        self.backbone_cat_tabular = backbone_continuous_tabular
+        # self.add_module('backbone', self.backbone)
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:157]: remove

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+        self.heads = torch.nn.ModuleList(heads)
+        self.add_module('heads', self.heads)
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:159]: remove

File: fuse_examples/classification/multimodality/model_tabular_imaging.py (hunk 10 above)
Comment (mosheraboh) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:148]: What about backbone_continuous_input?

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+                 tabular_heads: Sequence[torch.nn.Module]=None,
+                 imaging_heads: Sequence[torch.nn.Module]=None,
+                 multimodal_heads: Sequence[torch.nn.Module]=None,
+                 ) -> None:
+        super().__init__()
+
+        self.tabular_inputs = tabular_inputs
+        self.tabular_backbone = tabular_backbone
+        if self.tabular_backbone:
+            self.add_module('tabular_backbone', self.tabular_backbone)
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:190]: Why do you need the add_module?

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+
+#Tabular Imaging model
+class FuseMultiModalityModel(torch.nn.Module):
+    def __init__(self,
+                 tabular_inputs: Tuple[Tuple[str, int], ...]=None,
+                 imaging_inputs: Tuple[Tuple[str, int], ...]=None,
+                 tabular_backbone: torch.nn.Module=None,
+                 imaging_backbone: torch.nn.Module=None,
+                 multimodal_backbone: torch.nn.Module=None,
+                 tabular_heads: Sequence[torch.nn.Module]=None,
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:181]: Why do you care if it's imaging, tabular or multimodal heads?
If you don't care, you can maintain a single list for all heads.

File: fuse_examples/classification/multimodality/model_tabular_imaging.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+                batch_dict = head.forward(batch_dict)
+
+        if self.multimodal_heads:
+            for head in self.multimodal_heads:
+                batch_dict = head.forward(batch_dict)
+
+        return batch_dict['model']
+
+
+if __name__ == '__main__':
```
Comment (mosheraboh) [inline fuse_examples/classification/multimodality/model_tabular_imaging.py:254]: remove. add it as a unitest if it's important

**Review by mosheraboh — COMMENTED** [review]
Hi! Looks good.
The main question we should ask ourselves is what do we want to create here?
end to end example/template for multimodality /shared technology?
If all of them (which is my guess), what are the technologies that you want to create? can you separate it from the rest of the example and be extra pedantic with the explanations/type annotations for those components?
As you mentioned in our talk, we need to reorganize the dataset code both to be more simple and also to use fuse2 code.

File: fuse_examples/classification/multimodality/dataset.py [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+    augmentor = FuseAugmentorDefault(
+        augmentation_pipeline=augmentation_pipeline)
+
+
+
+
+    return augmentor
+
+
+def TABULAR_dataset(tabular_processor,df,tabular_features,sample_key):
```
[the remark this answers is not in the document]
Comment (taltlusty) [reply fuse_examples/classification/multimodality/dataset.py:73]: removed this func

File: fuse_examples/classification/multimodality/mg_dataset.py (hunk 3 above)
Comment (taltlusty) [reply fuse_examples/classification/multimodality/mg_dataset.py:40]: Yes, single value features

File: fuse_examples/classification/multimodality/model_tabular_imaging.py (hunk 7 above)
Comment (taltlusty) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:116]: added tabular fix

File: fuse_examples/classification/multimodality/model_tabular_imaging.py (hunk 10 above)
Comment (taltlusty) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:148]: didn't understand this comment - should I change the name to*_input

File: fuse_examples/classification/multimodality/model_tabular_imaging.py (hunk 15 above)
Comment (taltlusty) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:254]: will remove at the end of the review

File: fuse_examples/classification/multimodality/model_tabular_imaging.py (hunk 10 above)
Comment (mosheraboh) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:148]: No,
You have MLP for categorical_features and MLP for after concatenation.
Theoretically, you could add one more MLP  for the continuous features.
I guess my question is: do think it is interesting to add this option as well or not?

Comment (taltlusty) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:148]: added this option

File: fuse_examples/classification/multimodality/model_tabular_imaging.py (hunk 9 above)
Comment (taltlusty) [reply fuse_examples/classification/multimodality/model_tabular_imaging.py:64]: currently supports "current fuse" not fuse 2. will be changed
